sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoin.scala

```diff
@@ -62,10 +63,29 @@ case class BroadcastHashJoin(
   override def requiredChildDistribution: Seq[Distribution] =
     UnspecifiedDistribution :: UnspecifiedDistribution :: Nil
 
+  private[this] val broadcastFutureInitLock = new Object()
+
+  // Use @volatile so we can read a snapshot of this value of this without locking.
+  @volatile
+  private[this] var broadcastFutureValue: Option[Future[Broadcast[HashedRelation]]] = None
+
   // Use lazy so that we won't do broadcast when calling explain but still cache the broadcast value
   // for the same query.
```

Comment:
This comment needs to be updated.  Have you checked that the behavior described in the comment is maintined?